### PR TITLE
Adrien state of charge

### PR DESCRIPTION
**Pull request by acarrou**
The state of charge is up and running. I have tested and added a small demo that reads percentage and voltage. Now someone just needs to implement it. I purposely didn't work too much on the `IsBatteryLow()` function because I think this would be really good to give to someone new to learn how we could implement something like this into a rules engine. Note that this is using the max17043 device. If we had larger batteries we will need to edit this class so that we can calculate for just one cell but that's something we can look at later
state of charge
#67

### CODE REVIEW

File: common/state_of_charge.hpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    void Initialize()
+    {
+      module_.ModuleInitialize();
+    }
+
+    double BatteryPercentage()
+    {
+      units::voltage::volt_t battery_voltage = module_.GetVoltage();
+      double voltage_percent = double(battery_voltage);
```
Comment (naterpotatoers) [inline common/state_of_charge.hpp:21]:
```suggestion
      double voltage_percent = static_cast<double>(battery_voltage);
```
I think this is the safer c++ way

File: common/state_of_charge.hpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      voltage_percent = (voltage_percent - 3) / (4.2 - 3);
```
Comment (naterpotatoers) [inline common/state_of_charge.hpp:22]: Not sure where you got these numbers from? Do they mean something? Might want to create variables for them

File: common/state_of_charge.hpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      voltage_percent = voltage_percent * 100;
+      return voltage_percent;
+    }
+
+    float StateOfCharge()
+    {
+      units::voltage::volt_t battery_voltage = module_.GetVoltage();
+      return double(battery_voltage);
```
Comment (naterpotatoers) [inline common/state_of_charge.hpp:30]: use static_cast. Could also one line this e.g. return module.GetVoltage()

File: common/state_of_charge.hpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+      if (GetStateOfCharge() < 3.05)
+      {
+        sjsu::LogWarning("Cell Voltage below 3. Battery dangerously low!");
+        return true;
+      }
+      else if (GetStateOfCharge() < 3.25)
+      {
+        sjsu::LogWarning("Cell Voltage below 3.25. Battery is low");
+      }
```
Comment (naterpotatoers) [inline common/state_of_charge.hpp:53]: Duplicated? Ctrl + V one too many times smh

**Review by naterpotatoers — CHANGES_REQUESTED** [review]
Few potential problems detected

**Comment by acarrou** [comment]
> Few potential problems detected

Fixed.

File: common/state_of_charge.hpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      return voltage_percent;
+    }
+
+    float GetStateOfCharge()
+    {
+      return static_cast<float>(module_.GetVoltage());
+    }
+
+    bool IsBatteryLow()
+    {}
```
Comment (naterpotatoers) [inline common/state_of_charge.hpp:32]:
```suggestion
    { return false;}
```

File: common/state_of_charge.hpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  private:
+    const float full_battery_voltage = 4.2;
+    const float low_battery_voltage = 3.0;
```
Comment (naterpotatoers) [inline common/state_of_charge.hpp:36]: kCamelCase

**Review by naterpotatoers — APPROVED** [review]

File: common/state_of_charge.hpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      else if (GetStateOfCharge() < 3.25)
+      {
+        sjsu::LogWarning("Cell Voltage below 3.25. Battery is low");
+      }
+      return false;
+    }
+
+  private:
+    double full_battery_voltage = 4.2;
+    double low_battery_voltage = 3.05;
```
[the remark this answers is not in the document]
Comment (acarrou) [reply common/state_of_charge.hpp]: gotcha 👍

**Pull request merged**